magento/magento2
# Scheduled price rule time zone correction

### Description (*)

Apply the same logic for time zone correction on `\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule` as in `\Magento\CatalogRule\Model\Indexer\ReindexRuleProduct::execute`.

The problem here is that Magento 2.3 uses the method `\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule`. In Magento 2.4 this method is no longer used, but it is also not deprecated. So we can expect customers to use this method through the `@api` method `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexById`. So even though the code in this PR is not used anywhere in vanilla M2.4, it is an `@api` method that could result in catalog rules starting and ending at wrong time which is equal to the configured timezone offset.

Applying this PR makes sure that time zones are applied equally in `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexById` as they are in `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexByIds`. Currently they differ.

### Related Pull Requests


### Fixed Issues (if relevant)


### Manual testing scenarios (*)

1. Set time zone to Europe/Amsterdam
2. Set indexers to scheduled
3. Deleta all present Catalog Price Rules
4. Create a Catalog Price Rule for 50% off on all products, but disabled
5. Schedule the price rule to be enabled at a minute or so in the future using catalog staging
6. Wait until catalog stating rules are active and indexer is up to date
7. Run this query: `SELECT DISTINCT from_time, to_time FROM catalogrule_product;`
8. It should output only one row, after all, the rule is applied on all product at the same time.
9. Install [app.zip](https://github.com/magento/magento2/files/4868883/app.zip)
10. Run `bin/magento reindex:product 10`
11. Run this query: `SELECT DISTINCT from_time, to_time FROM catalogrule_product;`
12. Now you see two rows: one from before and one with a time offset 2 hours into the future. The Europe/Amsterdam timezone is not converted to UTC here. This is the issue.

### Questions or comments

`\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule` and `\Magento\CatalogRule\Model\Indexer\ReindexRuleProduct::execute` do contain duplicate code. Maybe, after approving this PR, it would be worth while to deprecate or merge some of the methods.

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#29549: Scheduled price rule time zone correction
